src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,323 @@
+                _connectionId = connectionId;
+                _invocationId = invocationId;
+                _token = cancellationToken;
+            }
+
+            // Needs to be called after adding the completion to the dictionary in order to avoid synchronous completions of the token registration
+            // not canceling when the dictionary hasn't been updated yet.
+            public void RegisterCancellation()
+            {
+                if (_token.CanBeCanceled)
+                {
+                    _tokenRegistration = _token.UnsafeRegister(static o =>
+                    {
+                        var tcs = (TaskCompletionSourceWithCancellation<T>)o!;
+                        tcs.SetCanceled();
+                    }, this);
+                }
+            }
+            public new void SetCanceled()
+            {
+                // TODO: RedisHubLifetimeManager will want to notify the other server (if there is one) about the cancellation
+                // so it can clean up state and potentially forward that info to the connection
+                _clientResultsManager.TryCompleteResult(_connectionId, CompletionMessage.WithError(_invocationId, "Canceled"));
+            }
+
+            public new void SetResult(T result)
+            {
+                _tokenRegistration.Dispose();
+                base.SetResult(result);
+            }
+
+            public new void SetException(Exception exception)
+            {
+                _tokenRegistration.Dispose();
+                base.SetException(exception);
+            }
+
+#pragma warning disable IDE0060 // Remove unused parameter
+            // Just making sure we don't accidentally call one of these without knowing
+            public static new void SetCanceled(CancellationToken cancellationToken) => Debug.Assert(false);
+
+            public static new void SetException(IEnumerable<Exception> exceptions) => Debug.Assert(false);
+            public static new bool TrySetCanceled()
+            {
+                Debug.Assert(false);
+                return false;
+            }
+            public static new bool TrySetCanceled(CancellationToken cancellationToken)
+            {
+                Debug.Assert(false);
+                return false;
+            }
+            public static new bool TrySetException(IEnumerable<Exception> exceptions)
+            {
+                Debug.Assert(false);
+                return false;
+            }
+            public static new bool TrySetException(Exception exception)
+            {
+                Debug.Assert(false);
+                return false;
+            }
+            public static new bool TrySetResult(T result)
+            {
+                Debug.Assert(false);
+                return false;
+            }
+#pragma warning restore IDE0060 // Remove unused parameter
+        }
+    }
+}
+#else
+namespace Microsoft.Azure.SignalR
+{ 
+    internal class ClientResultsManager: IClientResultsManager
```

Comment:
what's this?

Reply:
for .NET < 7.0, `IClientResultsManager` and `ClientResultsManager` are both empty